### DISCUSSION TITLE
How to obtain inference speed statistics?

### DISCUSSION THREAD

**Discussion by Mars-204** [ultralytics/ultralytics#8292] [Q&A]
I have trained a YOLOv8n model on custom dataset. For inference on a test dataset, I am using the model.val() function. But this gives a single value at the end of the validation. I want to obtain the inference statistics of each image validation so that I can plot the statistical mean, standard deviation and variance of inference speed. How can I obtain this data?

**Comment by glenn-jocher** [comment]
Hello! 👋 To gather detailed inference speed statistics for each image during validation, you can modify the validation loop to record the inference times. Here's a quick example using Python:

```python
import time
from ultralytics import YOLO

# Load your trained model
model = YOLO('path/to/your/yolov8n.pt')

# Prepare your dataset
data = 'path/to/your/dataset.yaml'

# Initialize lists to store inference times
inference_times = []

# Run validation
for img in model.val(data=data):
    start_time = time.time()
    _ = model.predict(img)
    end_time = time.time()
    
    # Calculate and store inference time for the current image
    inference_times.append(end_time - start_time)

# Now you can calculate mean, std, and variance using inference_times list
mean_inference_time = sum(inference_times) / len(inference_times)
std_inference_time = (sum([(x - mean_inference_time) ** 2 for x in inference_times]) / len(inference_times)) ** 0.5
variance_inference_time = std_inference_time ** 2

print(f"Mean Inference Time: {mean_inference_time}")
print(f"Standard Deviation of Inference Time: {std_inference_time}")
print(f"Variance of Inference Time: {variance_inference_time}")

# Now you can plot your statistics as needed
```

This code snippet will calculate the inference time for each image and store it. After running through all images, you can compute the mean, standard deviation, and variance. You can then use these statistics to create your plots. Happy coding! 😊

**Reply by Mars-204** [reply]
@glenn-jocher Thank you for the reply. Unfortunately when I run this I get the same output as just running model.val(). It doesn't go inside the for loop for further eval. I just get the below output, which is model.eval:
```
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 384/384 [00:44<00:00,  8.70it/s]
                   all       1536       1730      0.422     0.0983      0.119      0.066
Speed: 1.4ms preprocess, 12.8ms inference, 0.0ms loss, 4.0ms postprocess per image
```
I modified the approach by loading each image from dataset and the passing it through the model.predict(). As below:

```
        inference_times = []
        all_images = list(Path(r'D:\data\test').glob('*.npy'))
        for img in all_images:
            np_image = np.load(img)
            start_time = time.time()
            _ = self.model_predict.predict(source=np_image, imgsz=512, device=1)
            end_time = time.time()
```
But when I calculate the inference time I get mean inference time as 56.29 ms which is way higher than the inference time while using model.eval(12.8ms in above case). What is the difference between these and how to appraoch further

**Reply by glenn-jocher** [reply]
Hey there! 👋 It looks like you're on the right track with loading and processing each image individually to gather inference times. The discrepancy you're seeing between the `model.val()` and your looped `model.predict()` times could be due to several factors, including additional overhead from loading images and the `predict()` function handling each image separately rather than in batches.

To minimize this overhead and get more accurate inference times, you might want to preload all images into memory before starting your timing loop. Also, ensure that you're not including any pre-processing or post-processing time in your measurements if you're looking for pure inference speed. Here's a streamlined way to do it:

```python
from pathlib import Path
import numpy as np
import time

# Preload images
all_images = [np.load(img) for img in Path(r'D:\data\test').glob('*.npy')]

# Initialize list for inference times
inference_times = []

# Measure inference time
for np_image in all_images:
    start_time = time.time()
    _ = model.predict(source=np_image, imgsz=512, device=1)
    end_time = time.time()
    inference_times.append(end_time - start_time)

# Calculate statistics
mean_inference_time = sum(inference_times) / len(inference_times)
# ... calculate std and variance as before

print(f"Mean Inference Time: {mean_inference_time:.2f} ms")
```

This should give you a more accurate representation of the inference time. Keep in mind that `model.val()` is optimized for batch processing, which is why it's faster. If you have any more questions or need further assistance, feel free to ask. Happy coding! 😊

**Reply by Mars-204** [reply]
Thanks for the reply and detailed explaination.

**Reply by glenn-jocher** [reply]
You're welcome! If you have any more questions or need further assistance, don't hesitate to reach out. Happy coding! 😊🚀
